File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates, target, result, new ArrayList<>(), 0);
        return result;
    }
    
    public void backtrack(int[] candidates, int target, List<List<Integer>> result, List<Integer> current, int start){
        if(target == 0 && current.size() > 0){
            result.add(new ArrayList<>(current));
            return;
        } else if(target < 0){
            return;
        } else {
            for(int i=start;i<candidates.length;i++){
                current.add(candidates[i]);
                backtrack(candidates, target-candidates[i], result, current, i);
                current.remove(current.size()-1);
            }
            return;
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        def backtrack(current, remain, start):
            if remain == 0:
                result.append(current.copy())
            elif remain < 0:
                return
            else:
                for i in range(start, len(candidates)):
                    current.append(candidates[i])
                    backtrack(current, remain-candidates[i], i)
                    current.pop()
        backtrack([], target, 0)
        return result